pkg/cmd/cmd_update.go

```diff
@@ -48,39 +57,163 @@ func KpmUpdate(c *cli.Context, kpmcli *client.KpmClient) error {
+			var releases []string
+			if dep.Git != nil {
+				releases, err = mvs.GetReleasesFromSource(dep.GetSourceType(), dep.Git.Url)
+			} else if dep.Oci != nil {
+				releases, err = mvs.GetReleasesFromSource(dep.GetSourceType(), "fjdkasd")
+			}
+			if err != nil {
+				reporter.ReportEventToStdout(reporter.NewEvent(
+					reporter.FailedGetReleases,
+					fmt.Sprintf("failed to get releases for %s", pkgName),
+				))
+				continue
+			}
+			pkgVersion, err = semver.LatestCompatibleVersion(releases, dep.Version)
+			if err != nil {
+				reporter.NewErrorEvent(
+					reporter.FailedSelectLatestCompatibleVersion,
+					err,
+					fmt.Sprintf("failed to find the latest version for %s", pkgName),
+				)
+				continue
+			}
+		}
+		if pkgVersion < dep.Version {
+			modulesToDowngrade = append(modulesToDowngrade, module.Version{Path: pkgName, Version: pkgVersion})
+		} else if pkgVersion > dep.Version {
+			modulesToUpgrade = append(modulesToUpgrade, module.Version{Path: pkgName, Version: pkgVersion})
 		}
+	}
+
+	_, depGraph, err := kpmcli.InitGraphAndDownloadDeps(kclPkg)
+	if err != nil {
+		return err
+	}
+
+	reqs := mvs.ReqsGraph{
+		Graph:     depGraph,
+		KpmClient: kpmcli,
+		KpmPkg:    kclPkg,
+	}
+
+	target := module.Version{Path: kclPkg.GetPkgName(), Version: kclPkg.GetPkgVersion()}
+	buildList, err := UpdateBuildList(target, modulesToUpgrade, modulesToDowngrade, &reqs)
+	if err != nil {
+		return reporter.NewErrorEvent(reporter.FailedUpdatingBuildList, err, "failed to update build list")
+	}
+
+	modules, err := graph.TopologicalSort(depGraph)
+	if err != nil {
+		return reporter.NewErrorEvent(reporter.FailedTopologicalSort, err, "failed to sort the dependencies")
+	}
 
-		err = kpmcli.UpdateDeps(kclPkg)
+	kclPkg.ModFile.Dependencies.Deps = make(map[string]pkg.Dependency)
+
+	for _, module := range modules {
+		if module.Path == target.Path || !slices.Contains(buildList, module) {
+			continue
+		}
+		d := pkg.Dependency{
+			Name:    module.Path,
+			Version: module.Version,
+		}
+		d.FullName = d.GenDepFullName()
+		_, properties, err := depGraph.VertexWithProperties(module)
 		if err != nil {
-			return err
+			return reporter.NewErrorEvent(reporter.FailedGetVertexProperties, err, "failed to get vertex with properties")
+		}
+		// there must be one property depending on the download source type
+		for sourceType, uri := range properties.Attributes {
+			d.Source, err = pkg.GenSource(sourceType, uri, module.Version)
+			if err != nil {
+				return reporter.NewErrorEvent(reporter.FailedGenerateSource, err, "failed to generate source")
+			}
+			// if d.Oci != nil && d.Oci.Reg == "ghcr.io" {
+			// 	d.Oci = nil
+			// }
 		}
+		kclPkg.ModFile.Dependencies.Deps[module.Path] = d
+	}
+
+	err = kpmcli.UpdateDeps(kclPkg)
+	if err != nil {
+		return err
 	}
 	return nil
 }
+
+func UpdateBuildList(target module.Version, modulesToUpgrade []module.Version, modulesToDowngrade []module.Version, reqs *mvs.ReqsGraph) ([]module.Version, error) {
```

Comment:
Add more comments on this function and put it into another package instead of `cmd` package.

Reply:
done